librz/util/base16.c

```diff
@@ -0,0 +1,290 @@
+}
+
+/** \internal
+ * \brief Decode a single Base‑16 output group.
+ * \param[in] src The encoded group of 2 Base‑16 characters.
+ * \param[out] dest The decoded single binary byte.
+ *
+ * This function operates on exactly 2 Base‑16 symbols and reconstructs
+ * the original byte (2 × 4 bits = 8 bits = 1 byte).
+ */
+static void unpack_from4(ut8 dest[1], const ut8 src[2]) {
+	size_t high = cd16(src[0]);
+	size_t low = cd16(src[1]);
+	dest[0] = (high << 4) | low;
+}
+
+/** \internal
+ * \brief Validate a Base‑16 (hexadecimal) character.
+ * \param c  The character to examine (promoted to \c int by the caller).
+ *
+ * The function returns true if:
+ * - an upper‑case letter in the range <code>'A'–'F'</code>, or
+ * - a lower‑case letter in the range <code>'a'–'f'</code> (accepted for
+ *   leniency), or
+ * - a digit in the range <code>'0'–'9'</code>.
+ *
+ * Any other value yields \c false.
+ */
+static bool is_base16(int c) {
+	return (c >= '0' && c <= '9') ||
+		(c >= 'A' && c <= 'F') ||
+		(c >= 'a' && c <= 'f'); /* allow lower‑case */
+}
+
+/** \internal
+ * \brief Calculate the length of \p src.
+ * \param[in] src The binary data to be Base16-encoded later.
+ * \param len The length of the binary data in bytes.
+ *
+ * This function returns \p len as it is unless it is negative, in
+ * which case, it returns the string length of \p src. A possibility
+ * of a string size larger than \c ST64_MAX requires us to make a
+ * bounds check, and error if overflow is possible. This function is
+ * provided in lieu of modifying the decoding API parameter list.
+ */
+static st64 calculate_src_length(const char *src, st64 len) {
+	size_t real_len;
+	if (len < 0) {
+		real_len = strlen(src);
+		if (ST64_MAX < real_len) {
+			return -1;
+		}
+		len = (st64)real_len;
+	}
+	return len;
+}
+
+/** \internal
+ * \brief Calculate the length, in bytes, of the Base‑16 encoded result.
+ * \param      len the length of the binary input in bytes.
+ *
+ * Base‑16 encodes each byte as two hexadecimal characters. Therefore,
+ * the output length is always exactly twice the input length.
+ */
+static size_t calculate_dest_length(size_t len) {
+	return 2 * len;
+}
+
+/**
+ * \brief Base16‑encode binary data (hexadecimal encoding).
+ * \param[out] dest The encoded output.
+ * \param[in]  src  The binary input data.
+ * \param      n    The length of the binary data in bytes.
+ * \return The length of the encoded output, excluding the NUL byte.
+ * \attention The \p dest parameter must have at least \c 1 + 2 × n bytes.
+ *
+ * Base16 encoding writes two ASCII characters per byte of input using
+ * uppercase hexadecimal characters ‘0’–‘9’, ‘A’–‘F’. No padding is used.
+ *
+ * If \p dest or \p src is \c NULL, nothing is written and \c 0 is returned.
+ */
+RZ_API size_t rz_base16_encode(RZ_OUT RZ_NULLABLE char *dest, RZ_NULLABLE const ut8 *src, size_t n) {
+	ut8 group[1];
+	size_t ret;
+	rz_return_val_if_fail(src, 0);
+	rz_return_val_if_fail(dest, 0);
```

Comment:
```suggestion
	rz_return_val_if_fail(src && dest, 0);
```
And move this at line 127

Reply:
updated in 9a4203e4c7